core/iwasm/interpreter/wasm_interp_classic.c

```diff
@@ -1404,12 +1404,42 @@ wasm_interp_call_func_import(WASMModuleInstance *module_inst,
 #define HANDLE_OP(opcode) HANDLE_##opcode:
 #define FETCH_OPCODE_AND_DISPATCH() goto *handle_table[*frame_ip++]
 
+#if WASM_ENABLE_THREAD_MGR != 0 && WASM_ENABLE_DEBUG_INTERP != 0
+#define HANDLE_OP_END()                                                       \
+    do {                                                                      \
+        /* Record the current frame_ip, so when exception occurs,             \
+           debugger can know the exact opcode who caused the exception */     \
+        frame_ip_orig = frame_ip;                                             \
+        /* atomic load the signal_flag first, and handle more with lock       \
+           if it is WAMR_SIG_SINGSTEP */                                      \
+        if (BH_ATOMIC_32_LOAD(exec_env->current_status->signal_flag)          \
+            == WAMR_SIG_SINGSTEP) {                                           \
+            os_mutex_lock(&exec_env->wait_lock);                              \
+            while (exec_env->current_status->signal_flag == WAMR_SIG_SINGSTEP \
+                   && exec_env->current_status->step_count++ == 1) {          \
+                exec_env->current_status->step_count = 0;                     \
+                SYNC_ALL_TO_FRAME();                                          \
+                wasm_cluster_thread_waiting_run(exec_env);                    \
+            }                                                                 \
+            os_mutex_unlock(&exec_env->wait_lock);                            \
+        }                                                                     \
+        goto *handle_table[*frame_ip++];                                      \
+    } while (0)
+#else
+#define HANDLE_OP_END() FETCH_OPCODE_AND_DISPATCH()
+#endif
+
+#else /* else of WASM_ENABLE_LABELS_AS_VALUES */
+#define HANDLE_OP(opcode) case opcode:
 #if WASM_ENABLE_THREAD_MGR != 0 && WASM_ENABLE_DEBUG_INTERP != 0
 #define HANDLE_OP_END()                                                   \
-    do {                                                                  \
-        /* Record the current frame_ip, so when exception occurs,         \
-           debugger can know the exact opcode who caused the exception */ \
-        frame_ip_orig = frame_ip;                                         \
+    /* Record the current frame_ip, so when exception occurs,             \
+       debugger can know the exact opcode who caused the exception */     \
+    frame_ip_orig = frame_ip;                                             \
+    /* atomic load the signal_flag first, and handle more with lock       \
+       if it is WAMR_SIG_SINGSTEP */                                      \
```

Comment:
can you add a comment why you don't need to check BH_ATOMIC_32_IS_ATOMIC?

Reply:
IIUC, in most cpu architectures, the 32-bit load operation is atomic when the address to load is 4-byte aligned. But it should be safer to atomic load the signal_flag or load with lock according to BH_ATOMIC_32_IS_ATOMIC, I updated the code to handle BH_ATOMIC_32_IS_ATOMIC, please help review again. Since most cpu supports 32-bit atomic load, they should be able to benefit from the PR. And for cpu that doesn't support 32-bit atomic load, the performance should be almost the same since the second lock occurs only when the signal_flag gotten is WAMR_SIG_SINGSTEP.

Reply:
i'm not quite sure about the locking protocol here.
to me, it seems that the signal_flag is modified w/o any lock or atomic operations. (eg. wasm_cluster_send_signal_all)

Reply:
Yes, I didn't check that before, it is really chaotic, `wasm_cluster_clear_thread_signal` and `wasm_cluster_thread_send_signal` sets `signal_flag` without lock, and some of its callers call them with lock, e.g., `wasm_cluster_thread_continue` and `set_thread_cancel_flags`, some call them without lock, e.g., `wasm_cluster_send_signal_all`. I think we can merge this PR first and fix them later, I need to discuss with others for how to fix them.